packages/request/lib/Agent.js

```diff
@@ -0,0 +1,341 @@
+      } else {
+        setTimeout(removeSession, 100 - duration)
+      }
+    }
+  }
+
+  clearSessionTimers (session) {
+    if (session[kTimeoutId]) {
+      clearTimeout(session[kTimeoutId])
+      session[kTimeoutId] = undefined
+    }
+    if (session[kIntervalId]) {
+      clearInterval(session[kIntervalId])
+      session[kIntervalId] = undefined
+    }
+  }
+
+  replaceSessionRequest (session, options = {}) {
+    const {
+      timeout = this.timeout
+    } = options
+    const semaphore = session[kSemaphore]
+    session[kRequest] = session.request
+    session.request = async function request (headers, options) {
+      await session[kConnected]
+      clearTimeout(session[kTimeoutId])
+      const release = await semaphore.acquire()
+      const stream = session[kRequest](headers, options)
+      stream.once('error', err => {
+        if (!isAbortError(err)) {
+          logger.error('Request -> stream error: %s', err.message)
+        }
+      })
+      stream.once('close', () => {
+        release()
+        if (semaphore.value >= session.remoteSettings.maxConcurrentStreams) {
+          session[kTimeoutId] = setTimeout(() => session.close(), timeout)
+        }
+      })
+      return stream
+    }
+  }
+
+  setSessionHeartbeat (session, pingInterval = this.pingInterval) {
+    const ping = promisify(session.ping).bind(session)
+    let canceled = false
+    const cancel = err => {
+      this.clearSessionTimers(session)
+      this.deleteSession(session)
+      session.destroy(err, NGHTTP2_CANCEL)
+    }
+    session[kIntervalId] = setInterval(async () => {
+      try {
+        const duration = await ping()
+        logger.trace('Session -> ping %d ms', Math.round(duration))
+      } catch (err) {
+        if (!canceled) {
+          logger.debug('Session -> canceled: %s', err.message)
+          canceled = true
+          cancel(err)
+        }
+      }
+    }, pingInterval)
+  }
+
+  setSessionTimeout (session, timeout = this.timeout) {
+    session.setTimeout(timeout, () => {
+      logger.info('Session -> timeout')
+      this.clearSessionTimers(session)
+      this.deleteSession(session)
+      session.destroy()
+    })
+  }
+
+  addSessionListeners (session, options = {}) {
+    const {
+      connectTimeout = this.connectTimeout,
+      timeout = this.timeout,
+      pingInterval = this.pingInterval
+    } = options
+    const sid = session[kSid]
+    const semaphore = session[kSemaphore]
+    session.socket.on('session', tlsSession => {
+      this.tlsSessionCache.set(sid, tlsSession)
+    })
+    const setMaxConcurrency = () => {
+      semaphore.maxConcurrency = session.remoteSettings.maxConcurrentStreams
+    }
+    const deleteTlsSession = () => {
+      this.tlsSessionCache.delete(sid)
+    }
+    const destroySession = () => {
+      this.clearSessionTimers(session)
+      this.deleteSession(session)
+      session.destroy()
+    }
+    session[kConnected] = new Promise((resolve, reject) => {
```

Comment:
The code inside the Promise should only handle the fact if the session has been connected or not. The code which handles the subsequent action on an established connection should be move outside.